File: moe/tools/moe.generator.project/src/main/resources/org/moe/generator/project/templates/master-detail/java/ui/MasterViewController.java
package ${_PLACEHOLDER_PACKAGE_NAME}.ui;

import org.moe.natj.general.Pointer;
import org.moe.natj.general.ann.Mapped;
import org.moe.natj.general.ann.NInt;
import org.moe.natj.general.ann.Owned;
import org.moe.natj.general.ann.RegisterOnStartup;
import org.moe.natj.objc.ObjCRuntime;
import org.moe.natj.objc.SEL;
import org.moe.natj.objc.ann.ObjCClassName;
import org.moe.natj.objc.ann.Selector;
import org.moe.natj.objc.map.ObjCObjectMapper;

import java.util.ArrayList;
import java.util.Date;

import apple.foundation.NSArray;
import apple.foundation.NSIndexPath;
import apple.uikit.UIBarButtonItem;
import apple.uikit.UINavigationController;
import apple.uikit.UIStoryboardSegue;
import apple.uikit.UITableView;
import apple.uikit.UITableViewCell;
import apple.uikit.UITableViewController;
import apple.uikit.enums.UIBarButtonSystemItem;
import apple.uikit.enums.UITableViewCellEditingStyle;
import apple.uikit.enums.UITableViewRowAnimation;

@org.moe.natj.general.ann.Runtime(ObjCRuntime.class)
@ObjCClassName("MasterViewController")
@RegisterOnStartup
public class MasterViewController extends UITableViewController {

    @Owned
    @Selector("alloc")
    public static native MasterViewController alloc();

    @Selector("init")
    public native MasterViewController init();

    protected MasterViewController(Pointer peer) {
        super(peer);
    }

    private DetailViewController detailViewController;

    private ArrayList<String> objects = new ArrayList<>();

    @Override
    public void viewDidLoad() {
        super.viewDidLoad();
        // Do any additional setup after loading the view, typically from a nib.
        navigationItem().setLeftBarButtonItem(editButtonItem());

        UIBarButtonItem addButton = UIBarButtonItem.alloc()
                .initWithBarButtonSystemItemTargetAction(UIBarButtonSystemItem.Add, this, new SEL
                        ("insertNewObject:"));
        navigationItem().setRightBarButtonItem(addButton);
        detailViewController = (DetailViewController) ((UINavigationController)
                splitViewController().viewControllers().lastObject()).topViewController();
    }

    @Override
    public void viewWillAppear(boolean animated) {
        setClearsSelectionOnViewWillAppear(splitViewController().isCollapsed());
        super.viewWillAppear(animated);
    }

    @Selector("insertNewObject:")
    public void insertNewObject(Object sender) {
        objects.add(0, new Date().toString());
        NSIndexPath indexPath = NSIndexPath.indexPathForRowInSection(0, 0);
        tableView().insertRowsAtIndexPathsWithRowAnimation((NSArray<? extends NSIndexPath>)
                NSArray.arrayWithObject(indexPath), UITableViewRowAnimation.Automatic);
    }

    @Override
    public void prepareForSegueSender(UIStoryboardSegue segue, @Mapped(ObjCObjectMapper.class)
            Object sender) {
        if (segue.identifier().equals("showDetail")) {
            NSIndexPath indexPath = tableView().indexPathForSelectedRow();
            final String object = objects.get((int) indexPath.row());
            final DetailViewController controller = (DetailViewController) (
                    (UINavigationController) segue.destinationViewController()).topViewController();
            controller.setDetailItem(object);
            controller.navigationItem().setLeftBarButtonItem(splitViewController()
                    .displayModeButtonItem());
            controller.navigationItem().setLeftItemsSupplementBackButton(true);
        }
    }

    @Override
    public long numberOfSectionsInTableView(UITableView uiTableView) {
        return 1;
    }

    @Override
    public long tableViewNumberOfRowsInSection(UITableView uiTableView, @NInt long l) {
        return objects.size();
    }

    @Override
    public UITableViewCell tableViewCellForRowAtIndexPath(UITableView tableView, NSIndexPath
            indexPath) {
        final UITableViewCell cell = tableView.dequeueReusableCellWithIdentifierForIndexPath
                ("Cell", indexPath);

        final String object = objects.get((int) indexPath.row());
        cell.textLabel().setText(object);
        return cell;
    }

    @Override
    public boolean tableViewCanEditRowAtIndexPath(UITableView uiTableView, NSIndexPath
            nsIndexPath) {
        return true;
    }

    @Override
    public void tableViewCommitEditingStyleForRowAtIndexPath(UITableView tableView, @NInt long
            editingStyle, NSIndexPath indexPath) {
        if (editingStyle == UITableViewCellEditingStyle.Delete) {
            objects.remove((int) indexPath.row());
            tableView.deleteRowsAtIndexPathsWithRowAnimation((NSArray<? extends NSIndexPath>)
                    NSArray.arrayWithObject(indexPath), UITableViewRowAnimation.Fade);
        }
    }
}

File: moe/tools/moe.generator.project/src/main/resources/org/moe/generator/project/templates/master-detail/kotlin/ui/MasterViewController.kt
package ${_PLACEHOLDER_PACKAGE_NAME}.ui

import apple.foundation.NSArray
import apple.foundation.NSIndexPath
import apple.uikit.*
import apple.uikit.enums.UIBarButtonSystemItem
import apple.uikit.enums.UITableViewCellEditingStyle
import apple.uikit.enums.UITableViewRowAnimation
import org.moe.natj.general.Pointer
import org.moe.natj.general.ann.Mapped
import org.moe.natj.general.ann.NInt
import org.moe.natj.general.ann.Owned
import org.moe.natj.general.ann.RegisterOnStartup
import org.moe.natj.objc.ObjCRuntime
import org.moe.natj.objc.SEL
import org.moe.natj.objc.ann.ObjCClassName
import org.moe.natj.objc.ann.Selector
import org.moe.natj.objc.map.ObjCObjectMapper
import java.util.*
import ${_PLACEHOLDER_PACKAGE_NAME}.ui.DetailViewController

@org.moe.natj.general.ann.Runtime(ObjCRuntime::class)
@ObjCClassName("MasterViewController")
@RegisterOnStartup
class MasterViewController protected constructor(peer: Pointer) : UITableViewController(peer) {

    @Selector("init")
    override external fun init(): MasterViewController

    private var detailViewController: DetailViewController? = null

    private val objects = ArrayList<String>()

    override fun viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        navigationItem().setLeftBarButtonItem(editButtonItem())

        val addButton = UIBarButtonItem.alloc()
                .initWithBarButtonSystemItemTargetAction(UIBarButtonSystemItem.Add, this, SEL("insertNewObject:"))
        navigationItem().setRightBarButtonItem(addButton)
        detailViewController = (splitViewController().viewControllers().lastObject() as UINavigationController).topViewController() as DetailViewController
    }

    override fun viewWillAppear(animated: Boolean) {
        setClearsSelectionOnViewWillAppear(splitViewController().isCollapsed)
        super.viewWillAppear(animated)
    }

    @Selector("insertNewObject:")
    fun insertNewObject(sender: Any) {
        objects.add(0, Date().toString())
        val indexPath = NSIndexPath.indexPathForRowInSection(0, 0)
        tableView().insertRowsAtIndexPathsWithRowAnimation(NSArray.arrayWithObject(indexPath) as NSArray<out NSIndexPath>, UITableViewRowAnimation.Automatic)
    }

    override fun prepareForSegueSender(segue: UIStoryboardSegue, @Mapped(ObjCObjectMapper::class)
    sender: Any) {
        if (segue.identifier() == "showDetail") {
            val indexPath = tableView().indexPathForSelectedRow()
            val `object` = objects[indexPath.row().toInt()]
            val controller = (segue.destinationViewController() as UINavigationController).topViewController() as DetailViewController
            controller.detailItem = `object`
            controller.navigationItem().setLeftBarButtonItem(splitViewController()
                    .displayModeButtonItem())
            controller.navigationItem().setLeftItemsSupplementBackButton(true)
        }
    }

    override fun numberOfSectionsInTableView(uiTableView: UITableView?): Long {
        return 1
    }

    override fun tableViewNumberOfRowsInSection(uiTableView: UITableView, @NInt l: Long): Long {
        return objects.size.toLong()
    }

    override fun tableViewCellForRowAtIndexPath(tableView: UITableView, indexPath: NSIndexPath): UITableViewCell {
        val cell = tableView.dequeueReusableCellWithIdentifierForIndexPath("Cell", indexPath)

        val `object` = objects[indexPath.row().toInt()]
        cell.textLabel().setText(`object`)
        return cell
    }

    override fun tableViewCanEditRowAtIndexPath(uiTableView: UITableView?, nsIndexPath: NSIndexPath?): Boolean {
        return true
    }

    override fun tableViewCommitEditingStyleForRowAtIndexPath(tableView: UITableView?, @NInt editingStyle: Long, indexPath: NSIndexPath?) {
        if (editingStyle == UITableViewCellEditingStyle.Delete) {
            objects.removeAt(indexPath!!.row().toInt())
            tableView!!.deleteRowsAtIndexPathsWithRowAnimation(NSArray.arrayWithObject(indexPath) as NSArray<out NSIndexPath>, UITableViewRowAnimation.Fade)
        }
    }

    companion object {

        @Owned
        @Selector("alloc")
        @JvmStatic external fun alloc(): MasterViewController
    }
}
